pkg/kubelet/kubelet.go

```diff
@@ -2238,32 +2239,88 @@ func (kl *Kubelet) rejectPod(pod *api.Pod, reason, message string) {
 		Message: "Pod " + message})
 }
 
+func (kl *Kubelet) getNodeResources() api.ResourceList {
+	nodeResource, err := kl.cadvisor.MachineInfo()
+	if err != nil {
+		fmt.Errorf("Error when getting resource information from cAdvisor")
+	}
+	return api.ResourceList{
+		api.ResourceCPU:    *resource.NewMilliQuantity(int64(nodeResource.NumCores)*1000, resource.DecimalSI),
+		api.ResourceMemory: *resource.NewQuantity(int64(nodeResource.MemoryCapacity), resource.BinarySI),
+		api.ResourcePods:   *resource.NewQuantity(int64(kl.maxPods), resource.DecimalSI),
+	}
+}
+
+func (kl *Kubelet) newNodeWithLocateState() *api.Node {
+	node := &api.Node{ObjectMeta: api.ObjectMeta{Name: kl.nodeName, Labels: kl.nodeLabels},
+		Status: api.NodeStatus{Capacity: api.ResourceList{}, Allocatable: kl.getNodeResources()}}
+	return node
+}
+
+func (kl *Kubelet) getNode() *api.Node {
+	var node *api.Node
+	// Below is not a good implementation of getting *api.Node, but it works.
+	// RunGeneralPredicates() requires *api.Node, which is obtained as follows:
+	// 1. When kubelet is in standaloneMode, manufactures a *api.Node with kubelet information
+	// 2. Try to get *api.Node from kubelet's nodeInfo, if failed, we try 3.
+	// 3. Try to get *api.Node from apiserver
+	if kl.standaloneMode {
+		return kl.newNodeWithLocateState()
```

Comment:
Yeah. We can keep a cached and regularly updated copy of NodeStatus which can be used here.